scripts/conversion_script.py

```diff
@@ -0,0 +1,477 @@
+# number as used originally.
+STATUS_CODE_NAME = {number: name for name, number in STATUS_CODE.items()}
+# Path to the convert2rhel report json file.
+C2R_REPORT_FILE = "/var/log/convert2rhel/convert2rhel-pre-conversion.json"
+# Path to the convert2rhel report textual file.
+C2R_REPORT_TXT_FILE = "/var/log/convert2rhel/convert2rhel-pre-conversion.txt"
+# Path to the convert2rhel migration results file.
+C2R_MIGRATION_RESULTS_FILE = "/etc/migration-results"
+
+
+class RequiredFile(object):
+    """Holds data about files needed to download convert2rhel"""
+
+    def __init__(self, path="", host=""):
+        self.path = path
+        self.host = host
+        self.sha512_on_system = None
+        self.is_file_present = False
+
+
+class ProcessError(Exception):
+    """Custom exception to report errors during setup and run of conver2rhel"""
+
+    def __init__(self, message, report):
+        super(ProcessError, self).__init__(report)
+        self.message = message
+        self.report = report
+
+
+class OutputCollector(object):
+    """Wrapper class for script expected stdout"""
+
+    def __init__(self, status="", message="", report="", entries=None):
+        self.status = status
+        self.message = message
+        self.report = report
+        self.tasks_format_version = "1.0"
+        self.tasks_format_id = "oamg-format"
+        self.entries = entries
+        self.report_json = None
+
+    def to_dict(self):
+        # If we have entries, then we change report_json to be a dictionary
+        # with the needed values, otherwise, we leave it as `None` to be
+        # transformed to `null` in json.
+        if self.entries:
+            self.report_json = {
+                "tasks_format_version": self.tasks_format_version,
+                "tasks_format_id": self.tasks_format_id,
+                "entries": self.entries,
+            }
+
+        return {
+            "status": self.status,
+            "message": self.message,
+            "report": self.report,
+            "report_json": self.report_json,
+        }
+
+
+def find_highest_report_level(actions):
+    """
+    Gather status codes from messages and result. We are not seeking for
+    differences between them as we want all the results, no matter from where
+    they come.
+    """
+    print("Collecting and combining report status.")
+    action_level_combined = []
+    for value in actions.values():
+        action_level_combined.append(value["result"]["level"])
+        for message in value["messages"]:
+            action_level_combined.append(message["level"])
+
+    valid_action_levels = [
+        level for level in action_level_combined if level in STATUS_CODE
+    ]
+    valid_action_levels.sort(key=lambda status: STATUS_CODE[status], reverse=True)
+    return valid_action_levels[0]
+
+
+def gather_json_report():
+    """Collect the json report generated by convert2rhel."""
+    print("Collecting JSON report.")
+    with open(C2R_REPORT_FILE, "r") as handler:
```

Comment:
Do we expect the file to exist? I mean is it required that the pre-conversion was run? To me it would make sense to officially require it and possibly include message in error that they can run the task before this conversion task.

Reply:
So, we will have a report json file in the end anyway, both with analysis and conversion. In our workflow, when we reach the *point of no return*, which is a special part of the script that should decide if:
1. we should rollback and stop the execution (analysis)
2. continue with the conversion and print the report (convert)

So, if we have the report, it means that the conversion went through most steps, either if succeeded or not.

Reply:
Sure, makes sense, let's go with that